File: src/Preparation.php
<?php
declare(strict_types=1);
namespace MTX;

/** Directory-backed preparation queue. Publishing remains an explicit action. */
final class Preparation
{
    public function __construct(private readonly App $app) {}
    public function claim(): ?array
    {
        return $this->app->store->change(function (&$s) {
            foreach ($s['releases'] as $id=>&$r) {
                if ($r['state']!=='draft' || $r['artifact_format']!=='prepared-payload-v1') continue;
                $p=$r['preparation']??['status'=>'queued','attempts'=>0];
                if ($p['status']==='failed' || ($p['status']==='processing' && ($p['started_at']??0)>time()-300)) continue;
                if (($p['attempts']??0)>=3) { $r['preparation']=$p+['error'=>'处理多次中断，请重试。']; $r['preparation']['status']='failed'; continue; }
                $token=bin2hex(random_bytes(16));
                $r['preparation']=['status'=>'processing','attempts'=>($p['attempts']??0)+1,'started_at'=>time(),'token'=>$token];
                return array_intersect_key($r,array_flip(['release_id','app_key','bundle_id','display_version','min_ios','source_sha256','source_bytes']))+
                    ['token'=>$token,'source_path'=>$this->app->object($r['source_sha256']),'work_root'=>$this->app->storage.'/preparation'];
            }
            return null;
        });
    }
    private function active(array $s,string $id,string $token): array
    {
        $r=Releases::release($s,$id);
        if ($r['state']!=='draft' || ($r['preparation']['status']??'')!=='processing' || !hash_equals($r['preparation']['token']??'', $token)) throw new Problem(409,'自动处理任务已失效。');
        return $r;
    }
    public function finish(string $id,string $token,string $path): void
    {
        $r=$this->active($this->app->store->read(),$id,$token);
        $expected=$this->app->storage.'/preparation/'.$id.'-'.$token.'/payload.tar';
        if ($path!==$expected || is_link($path) || !is_file($path) || realpath($path)!==$expected) throw new Problem(422,'自动处理输出路径异常。');
        $manifest=Packages::prepared($path,$r);
        $object=$this->app->saveObject($path);
        $this->app->store->change(function (&$s) use ($id,$token,$manifest,$object) {
            $this->active($s,$id,$token);
            $s['releases'][$id]['artifact_sha256']=$object['sha256'];
            $s['releases'][$id]['artifact_bytes']=$object['bytes'];
            $s['releases'][$id]['manifest_sha256']=$manifest;
            $s['releases'][$id]['state']='ready';
            $s['releases'][$id]['preparation']=['status'=>'complete','finished_at'=>time()];
            Store::audit($s,'TIPA 自动处理与校验完成',$id);
        });
    }
    public function fail(string $id,string $token,string $message): void
    {
        $message=Packages::label($message,200);
        $this->app->store->change(function (&$s) use ($id,$token,$message) {
            $r=$this->active($s,$id,$token);
            $s['releases'][$id]['preparation']=['status'=>'failed','error'=>$message,'attempts'=>$r['preparation']['attempts'],'finished_at'=>time()];
            Store::audit($s,'TIPA 自动处理未完成',$id);
        });
    }
    public function retry(string $key,string $id): void
    {
        $this->app->store->change(function (&$s) use ($key,$id) {
            $r=Releases::release($s,$id,$key);
            if ($r['state']!=='draft' || ($r['preparation']['status']??'')!=='failed') throw new Problem(409,'当前记录无需重试。');
            $s['releases'][$id]['preparation']=['status'=>'queued','attempts'=>0];
            Store::audit($s,'重新排队自动处理',$id);
        });
    }
}

File: tests/preparation.py
#!/usr/bin/env python3
"""Queue regression with disposable state; a real TIPA is read, never run."""
import argparse, hashlib, json, os, secrets, subprocess, tempfile, time
from pathlib import Path
ROOT=Path(__file__).resolve().parents[1];PHP=os.environ.get('PHP_BINARY','php')
p=argparse.ArgumentParser();p.add_argument('--source',required=True);a=p.parse_args();source=Path(a.source).resolve();checks=0

def check(ok,label):
 global checks
 assert ok,label
 checks+=1; print('PASS',label,flush=True)

with tempfile.TemporaryDirectory(prefix='mtx-auto-prepare-') as td:
 t=Path(td);config=t/'config.php';storage=t/'storage';env=os.environ|{'MTX_CONFIG':str(config)}
 subprocess.run([PHP,str(ROOT/'bin/setup.php'),'--url','http://127.0.0.1:19999','--local-http','--config',str(config),'--storage',str(storage),'--password-stdin'],input=secrets.token_hex(20)+'\n',text=True,check=True,capture_output=True)
 storage=storage.resolve()
 def php(code,*args):
  return subprocess.run([PHP,'-r','require "vendor/autoload.php";'+code,*map(str,args)],env=env,cwd=ROOT,text=True,capture_output=True,check=True).stdout
 def queue(data,ok=True):
  r=subprocess.run([PHP,str(ROOT/'bin/preparation-queue.php')],input=json.dumps(data),env=env,cwd=ROOT,text=True,capture_output=True)
  check(r.returncode==0 if ok else r.returncode!=0,'queue '+data['action']+(' accepted' if ok else 'rejected'))
  return json.loads(r.stdout) if r.returncode==0 else None
 def read():return json.loads((storage/'state.json').read_text())
 def write(s):(storage/'state.json').write_text(json.dumps(s))
 rid=php('$a=new MTX\\App();echo (new MTX\\Releases($a))->upload("mtx-dfm-cn",$argv[1],[]);',source)
 check(read()['releases'][rid]['preparation']['status']=='queued','single TIPA queues preparation')
 check('path' not in queue({'action':'location'}) and queue({'action':'location'})['work_root']==str(storage/'preparation'),'custom storage honored')
 job=queue({'action':'claim'});check(job['release_id']==rid,'claim expected record')
 check(queue({'action':'claim'}) is None,'concurrent worker does not duplicate lease')
 queue({'action':'finish','release_id':rid,'token':'0'*32,'path':'/tmp/wrong'},False)
 queue({'action':'finish','release_id':rid,'token':job['token'],'path':'/tmp/wrong'},False)
 s=read();s['releases'][rid]['preparation']['started_at']=int(time.time())-400;write(s)
 second=queue({'action':'claim'});check(second['token']!=job['token'],'expired lease receives new token')
 queue({'action':'fail','release_id':rid,'token':job['token'],'message':'obsolete'},False)
 queue({'action':'fail','release_id':rid,'token':second['token'],'message':'fixture error'})
 check(queue({'action':'claim'}) is None,'failed job waits for explicit retry')
 php('(new MTX\\Preparation(new MTX\\App()))->retry("mtx-dfm-cn",$argv[1]);',rid)
 check(read()['releases'][rid]['preparation']['status']=='queued','retry preserves upload without another file')
 result=subprocess.run(['python3',str(ROOT/'preparer/worker.py'),'--php',PHP],env=env,cwd=ROOT,text=True,capture_output=True,timeout=130)
 check(result.returncode==0,'worker runs with custom storage: '+result.stderr[:100])
 r=read()['releases'][rid]
 check(r['state']=='ready','single real TIPA becomes ready: '+str(r.get('preparation')))
 check(r['source_sha256']==hashlib.sha256(source.read_bytes()).hexdigest(),'original uploaded bytes preserved')
 check(r['sequence'] is None and read()['apps']['mtx-dfm-cn']['current_release_id'] is None,'preparation does not publish')
 check(not list((storage/'preparation').glob('*-*')),'private job temporary files removed')
 php('$a=new MTX\\App();MTX\\Packages::prepared($a->object($argv[1]),$a->store->read()["releases"][$argv[2]]);',r['artifact_sha256'],rid)
 check(True,'generated TAR matches existing native contract')
 # Late completions and historical records must never become live implicitly.
 s=read();s['releases'][rid]['state']='draft';s['releases'][rid]['preparation']={'status':'queued','attempts':0};write(s)
 job=queue({'action':'claim'});s=read();s['releases'][rid]['state']='withdrawn';write(s)
 queue({'action':'fail','release_id':rid,'token':job['token'],'message':'late'},False)
 check(read()['releases'][rid]['state']=='withdrawn','withdrawn record stays withdrawn')
 # Existing drafts without preparation metadata are picked up unchanged.
 s=read();s['releases'][rid]['state']='draft';s['releases'][rid].pop('preparation',None);write(s)
 job=queue({'action':'claim'});check(job['release_id']==rid,'historical source-only draft queues automatically')
 # Failed attempts never become published, and repeated interruption is bounded.
 s=read();s['releases'][rid]['preparation']['attempts']=3;s['releases'][rid]['preparation']['started_at']=int(time.time())-400;write(s)
 check(queue({'action':'claim'}) is None and read()['releases'][rid]['preparation']['status']=='failed','three interrupted attempts stop for manual retry')
 php('(new MTX\\Preparation(new MTX\\App()))->retry("mtx-dfm-cn",$argv[1]);',rid)
 # Corrupt only the disposable stored fixture, not the supplied source.
 obj=storage/'objects'/r['source_sha256'];saved=obj.read_bytes();obj.write_bytes(saved[:-1])
 result=subprocess.run(['python3',str(ROOT/'preparer/worker.py'),'--php',PHP],env=env,cwd=ROOT,text=True,capture_output=True,timeout=130)
 failed=read()['releases'][rid]
 check(failed['state']=='draft' and failed['preparation']['status']=='failed','bad source digest stops before native processing')
 check(not list((storage/'preparation').glob('*-*')),'failed work directory removed')
 obj.write_bytes(saved)
 php('(new MTX\\Preparation(new MTX\\App()))->retry("mtx-dfm-cn",$argv[1]);',rid)
 job=queue({'action':'claim'});work=storage/'preparation'/(rid+'-'+job['token']);work.mkdir(mode=0o700)
 output=work/'payload.tar';output.symlink_to(storage/'objects'/r['artifact_sha256'])
 queue({'action':'finish','release_id':rid,'token':job['token'],'path':str(output)},False)
 check(read()['releases'][rid]['state']=='draft','symlink output is rejected without changing state')
 output.unlink();output.write_bytes(b'invalid tar')
 queue({'action':'finish','release_id':rid,'token':job['token'],'path':str(output)},False)
 check(read()['releases'][rid]['state']=='draft','invalid prepared output stays unpublished')
 s=read();s['releases'][rid]['state']='withdrawn';write(s)
 output.write_bytes((storage/'objects'/r['artifact_sha256']).read_bytes())
 queue({'action':'finish','release_id':rid,'token':job['token'],'path':str(output)},False)
 check(read()['releases'][rid]['state']=='withdrawn','late valid completion never resurrects withdrawn release')
print(f'{checks} automatic preparation checks passed; all temporary state removed.')
